src/common/hist_util.cu

```diff
@@ -170,31 +171,32 @@ void ExtractCuts(int device, Span<SketchEntry> cuts,
  * \param weights_scan          Inclusive scan of weights for each entry in sorted_data.
  * \param column_sizes_scan     Describes the boundaries of column segments in sorted data.
  */
-void ExtractWeightedCuts(int device, Span<SketchEntry> cuts,
-                         size_t num_cuts_per_feature, Span<Entry> sorted_data,
+void ExtractWeightedCuts(int device,
+                         size_t num_cuts_per_feature,
+                         Span<Entry> sorted_data,
                          Span<float> weights_scan,
-                         Span<size_t> column_sizes_scan) {
+                         Span<size_t> column_sizes_scan,
+                         Span<SketchEntry> cuts) {
   dh::LaunchN(device, cuts.size(), [=] __device__(size_t idx) {
     // Each thread is responsible for obtaining one cut from the sorted input
     size_t column_idx = idx / num_cuts_per_feature;
     size_t column_size =
         column_sizes_scan[column_idx + 1] - column_sizes_scan[column_idx];
     size_t num_available_cuts =
-        min(size_t(num_cuts_per_feature), column_size);
+        min(static_cast<size_t>(num_cuts_per_feature), column_size);
     size_t cut_idx = idx % num_cuts_per_feature;
     if (cut_idx >= num_available_cuts) return;
-
     Span<Entry> column_entries =
         sorted_data.subspan(column_sizes_scan[column_idx], column_size);
-    Span<float> column_weights =
-        weights_scan.subspan(column_sizes_scan[column_idx], column_size);
 
-    float total_column_weight = column_weights.back();
+    Span<float> column_weights_scan =
+        weights_scan.subspan(column_sizes_scan[column_idx], column_size);
+    float total_column_weight = column_weights_scan.back();
     size_t sample_idx = 0;
     if (cut_idx == 0) {
       // First cut
       sample_idx = 0;
-    } else if (cut_idx == num_available_cuts - 1) {
+    } else if (cut_idx % num_available_cuts == 0) {
```

Comment:
This makes last cut for all feature special, instead of only the last cut of whole dataset.

Reply:
Doesn't cut_idx start from 0 again for each feature already?